Incorrect completion for labels with closed delimiters

**Describe the bug**
Incorrect completion when completing a label containing a syntactically valid label. This happens then when youre trying to get completions for a label while the delimiter is already closed `<...|>` with `|` being the cursor position.

![image](https://github.com/Myriad-Dreamin/tinymist/assets/137803093/7a1511b9-f615-4abb-82e9-3e5f8455c499)

Th following example works as expected:

![image](https://github.com/Myriad-Dreamin/tinymist/assets/137803093/825f40ff-cc47-4e73-a36c-109cd7f19696)

Since ref syntax `@...` does not have a closing delimiter this is not a problem in markup either.


**Package/Software version:**
tinymist version:
```plain
tinymist
Build Timestamp:     2024-04-17T12:01:31.978359992Z
Build Git Describe:  879d95a
Commit SHA:          879d95a74b19b91d1c1b183d87f4c0ada4bd4e75
Commit Date:         None
Commit Branch:       None
Cargo Target Triple: x86_64-unknown-linux-gnu
Typst Version:       0.11.0
```

**Logs:**
```plain
2024-05-17T14:38:17.004 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::harness] handling textDocument/completion - (113) at Instant { tv_sec: 18880, tv_nsec: 659562427 }\n"
2024-05-17T14:38:17.004 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::upstream::complete] continue after completing literal\n"
2024-05-17T14:38:17.004 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::global] def_use_lexical_hierarchy(/src/de/chapters/2-basics.typ): compute\n"
2024-05-17T14:38:17.005 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 681.043µs\n"
2024-05-17T14:38:17.005 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::global] import(223): compute\n"
2024-05-17T14:38:17.005 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::global] def_use((223, 2)): compute\n"
2024-05-17T14:38:17.005 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::def_use] diving source for def use: /src/util.typ\n"
2024-05-17T14:38:17.006 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::harness] handled  textDocument/completion - (113) in 1.50ms\n"
2024-05-17T14:38:17.326 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::server::lsp] notifying textDocument/didChange - at Instant { tv_sec: 18880, tv_nsec: 981301989 }\n"
2024-05-17T14:38:17.326 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::server::lsp] notifing textDocument/didChange succeeded in 227.56µs\n"
2024-05-17T14:38:17.326 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: compiling ...\n"
2024-05-17T14:38:17.326 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:17.333 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::harness] handling textDocument/completion - (114) at Instant { tv_sec: 18880, tv_nsec: 988136760 }\n"
2024-05-17T14:38:17.337 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: Compilation failed after 10.991129ms\n"
2024-05-17T14:38:17.337 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:17.337 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::actor::cluster] received diagnostics from primary: diag(Some(1))\n"
2024-05-17T14:38:17.344 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::actor::typ_server] CompileServerActor: evict compilation cache in 7.137341ms\n"
2024-05-17T14:38:17.345 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::upstream::complete] continue after completing literal\n"
2024-05-17T14:38:17.345 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::global] def_use_lexical_hierarchy(/src/de/chapters/2-basics.typ): compute\n"
2024-05-17T14:38:17.346 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 844.17µs\n"
2024-05-17T14:38:17.346 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::global] import(223): compute\n"
2024-05-17T14:38:17.346 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::global] def_use((223, 2)): compute\n"
2024-05-17T14:38:17.346 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist_query::analysis::def_use] diving source for def use: /src/util.typ\n"
2024-05-17T14:38:17.346 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:17Z INFO  tinymist::harness] handled  textDocument/completion - (114) in 13.84ms\n"
2024-05-17T14:38:18.021 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::server::lsp] notifying textDocument/didChange - at Instant { tv_sec: 18881, tv_nsec: 676837565 }\n"
2024-05-17T14:38:18.022 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::server::lsp] notifing textDocument/didChange succeeded in 226.35µs\n"
2024-05-17T14:38:18.022 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: compiling ...\n"
2024-05-17T14:38:18.022 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:18.088 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: Compilation failed after 65.938177ms\n"
2024-05-17T14:38:18.088 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:18.088 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::actor::cluster] received diagnostics from primary: diag(Some(1))\n"
2024-05-17T14:38:18.092 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::actor::typ_server] CompileServerActor: evict compilation cache in 3.962972ms\n"
2024-05-17T14:38:18.582 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::server::lsp] notifying textDocument/didChange - at Instant { tv_sec: 18882, tv_nsec: 237685691 }\n"
2024-05-17T14:38:18.583 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::server::lsp] notifing textDocument/didChange succeeded in 542.30µs\n"
2024-05-17T14:38:18.590 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: compiling ...\n"
2024-05-17T14:38:18.590 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:18.594 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: Compilation failed after 10.892829ms\n"
2024-05-17T14:38:18.594 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:18.594 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::actor::cluster] received diagnostics from primary: diag(Some(1))\n"
2024-05-17T14:38:18.601 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::actor::typ_server] CompileServerActor: evict compilation cache in 7.359998ms\n"
2024-05-17T14:38:18.870 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::server::lsp] notifying textDocument/didChange - at Instant { tv_sec: 18882, tv_nsec: 525712501 }\n"
2024-05-17T14:38:18.871 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::server::lsp] notifing textDocument/didChange succeeded in 578.98µs\n"
2024-05-17T14:38:18.871 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: compiling ...\n"
2024-05-17T14:38:18.871 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:18.881 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: Compilation failed after 9.713041ms\n"
2024-05-17T14:38:18.881 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:18.881 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::actor::cluster] received diagnostics from primary: diag(Some(1))\n"
2024-05-17T14:38:18.885 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:18Z INFO  tinymist::actor::typ_server] CompileServerActor: evict compilation cache in 3.952945ms\n"
2024-05-17T14:38:19.502 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  tinymist::server::lsp] notifying textDocument/didChange - at Instant { tv_sec: 18883, tv_nsec: 157244476 }\n"
2024-05-17T14:38:19.502 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  tinymist::server::lsp] notifing textDocument/didChange succeeded in 503.39µs\n"
2024-05-17T14:38:19.502 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: compiling ...\n"
2024-05-17T14:38:19.502 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:19.509 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  tinymist::harness] handling textDocument/completion - (115) at Instant { tv_sec: 18883, tv_nsec: 164150930 }\n"
2024-05-17T14:38:19.517 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  typst_ts_compiler::service::diag::console] /src/de/thesis.typ: Compilation failed after 14.548562ms\n"
2024-05-17T14:38:19.517 helix_lsp::transport [ERROR] tinymist err <- "    \n"
2024-05-17T14:38:19.517 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  tinymist::actor::cluster] received diagnostics from primary: diag(Some(1))\n"
2024-05-17T14:38:19.523 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  tinymist::actor::typ_server] CompileServerActor: evict compilation cache in 5.940497ms\n"
2024-05-17T14:38:19.524 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  tinymist_query::upstream::complete] continue after completing literal\n"
2024-05-17T14:38:19.524 helix_lsp::transport [ERROR] tinymist err <- "[2024-05-17T12:38:19Z INFO  tinymist::harness] handled  textDocument/completion - (115) in 15.22ms\n"
```

--- a/crates/tinymist-query/src/upstream/complete.rs
+++ b/crates/tinymist-query/src/upstream/complete.rs
@@ -40,10 +40,10 @@ pub fn autocomplete(
     mut ctx: CompletionContext,
 ) -> Option<(usize, bool, Vec<Completion>, Vec<lsp_types::CompletionItem>)> {
     let _ = complete_comments(&mut ctx)
+        || complete_labels(&mut ctx)
         || complete_type(&mut ctx).is_none() && {
             log::info!("continue after completing type");
             complete_field_accesses(&mut ctx)
-                || complete_open_labels(&mut ctx)
                 || complete_imports(&mut ctx)
                 || complete_rules(&mut ctx)
                 || complete_params(&mut ctx)
@@ -480,10 +480,12 @@ fn field_access_completions(ctx: &mut CompletionContext, value: &Value, styles: 
     }
 }
 
-/// Complete half-finished labels.
-fn complete_open_labels(ctx: &mut CompletionContext) -> bool {
+/// Complete labels.
+fn complete_labels(ctx: &mut CompletionContext) -> bool {
     // A label anywhere in code: "(<la|".
-    if ctx.leaf.kind().is_error() && ctx.leaf.text().starts_with('<') {
+    if (ctx.leaf.kind().is_error() && ctx.leaf.text().starts_with('<'))
+        || ctx.leaf.kind() == SyntaxKind::Label
+    {
         ctx.from = ctx.leaf.offset() + 1;
         ctx.label_completions();
         return true;

--- a/crates/tinymist/src/server/lsp_init.rs
+++ b/crates/tinymist/src/server/lsp_init.rs
@@ -368,11 +368,12 @@ impl Init {
                 definition_provider: Some(OneOf::Left(true)),
                 references_provider: Some(OneOf::Left(true)),
                 completion_provider: Some(CompletionOptions {
-                    // Please update the language-configurations.json if you are changing this
+                    // Please update the language-configuration.json if you are changing this
                     // setting.
                     trigger_characters: Some(vec![
                         String::from("#"),
                         String::from("("),
+                        String::from("<"),
                         String::from(","),
                         String::from("."),
                         String::from(":"),

--- a/tests/e2e/main.rs
+++ b/tests/e2e/main.rs
@@ -374,7 +374,7 @@ fn e2e() {
         });
 
         let hash = replay_log(&tinymist_binary, &root.join("neovim"));
-        insta::assert_snapshot!(hash, @"siphash128_13:6169526fa530f5a2d7907ffd23d67466");
+        insta::assert_snapshot!(hash, @"siphash128_13:d2a8dcc163ddbb5e204818aeb4c21b66");
     }
 
     {
@@ -385,7 +385,7 @@ fn e2e() {
         });
 
         let hash = replay_log(&tinymist_binary, &root.join("vscode"));
-        insta::assert_snapshot!(hash, @"siphash128_13:48f876724231ca7cdbadce094856da60");
+        insta::assert_snapshot!(hash, @"siphash128_13:32c954951f57dc40af56d3557fb15e5e");
     }
 }
 

